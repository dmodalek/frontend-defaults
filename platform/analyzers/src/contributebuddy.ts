import { directoryExists, getDependencyInstallation, DependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ContributebuddyAnalyzerResult = {
    contributebuddy: boolean;
    contributebuddyConfigFolder?: string,
    contributebuddyInstallation?: DependencyInstallation
};

export const contributebuddyAnalyzer = async (cwd: string): Promise<ContributebuddyAnalyzerResult> => {
    const CONTRIBUTE_BUDDY_FOLDER = join(cwd, '.contributebuddy');
    const contributebuddyConfigFolder = await directoryExists(CONTRIBUTE_BUDDY_FOLDER);
    const contributebuddyInstallation = await getDependencyInstallation(cwd, 'contribute-buddy');

    if (contributebuddyConfigFolder || contributebuddyInstallation.declared) {
        return {
            contributebuddy: true,
            contributebuddyConfigFolder: CONTRIBUTE_BUDDY_FOLDER,
            contributebuddyInstallation

        }
    }

    return {
        contributebuddy: false,
    };
}