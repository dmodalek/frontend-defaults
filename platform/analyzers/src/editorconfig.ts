import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type EditorconfigAnalyzerResult = {
	editorConfig: boolean;
};

export class EditorconfigAnalyzer extends Analyzer<EditorconfigAnalyzerResult> {
	async analyze(): Promise<EditorconfigAnalyzerResult> {
		const eidtorConfigExists = await fileExists(join(this.context, '.editorconfig'));

		return {
			editorConfig: eidtorConfigExists,
		};
	}
}
