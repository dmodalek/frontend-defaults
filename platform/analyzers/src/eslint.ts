import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: DependencyInstallation
};

export const eslintAnalyzer = async (cwd: string): Promise<ESLintAnalyzerResult> => {
	const doesESLintExist = await fileExists(join(cwd, '.eslintrc.js'));

	if (doesESLintExist) {
		const doesESLintIgnoreExist = await fileExists(join(cwd, '.eslintignore'));
		const eslintInstallation = await getDependencyInstallation(cwd, 'eslint');

		return {
			eslint: doesESLintExist,
			eslintIgnore: doesESLintIgnoreExist,
			eslintInstallation
		};
	}

	return {
		eslint: doesESLintExist,
	};
}