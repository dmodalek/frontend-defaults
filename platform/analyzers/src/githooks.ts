import {
	DependencyInstallation,
	fileExists,
	getDependencyInstallation,
	getPackageJSON,
	IPackage,
} from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitHooksLocation = 'rc' | 'package';

export type GitHooksAnalyzerResult = {
	gitHooks: boolean;
	gitHooksInstallation?: DependencyInstallation;
	gitHooksLocation?: GitHooksLocation;
};

type PackageWithHuskyConfig = IPackage & {
	husky: any;
};

export const gitHooksAnalyzer = async (cwd: string): Promise<GitHooksAnalyzerResult> => {
	const doesGitHooksConfigExist = await fileExists(join(cwd, '.huskyrc'));
	const gitHooksInstallation = await getDependencyInstallation(cwd, 'husky');
	const pkg: PackageWithHuskyConfig = (await getPackageJSON(cwd)) as any;

	if (doesGitHooksConfigExist || pkg.husky) {
		return {
			gitHooks: doesGitHooksConfigExist,
			gitHooksLocation: pkg.husky ? 'package' : 'rc',
			gitHooksInstallation,
		};
	}

	return {
		gitHooks: false,
	};
};
