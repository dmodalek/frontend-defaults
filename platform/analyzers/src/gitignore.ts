import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export const gitIgnoreAnalyzer = async (cwd: string): Promise<GitIgnoreAnalyzerResult> => {
	const doesGitignoreExist = await fileExists(join(cwd, '.gitignore'));

	return {
		gitignore: doesGitignoreExist,
	};
}