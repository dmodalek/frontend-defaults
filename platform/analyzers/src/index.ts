// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
