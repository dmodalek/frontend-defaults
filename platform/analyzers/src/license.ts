import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type LicenseAnalyzerResult = {
	license: boolean;
};

export class LicenseAnalyzer extends Analyzer<LicenseAnalyzerResult> {
	async analyze(): Promise<LicenseAnalyzerResult> {
		const licenseExists = await fileExists(join(this.context, 'LICENSE'));

		return {
			license: licenseExists,
		};
	}
}
