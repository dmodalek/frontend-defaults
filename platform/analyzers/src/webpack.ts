import {
	DependencyInstallation,
	findFilesByPattern,
	getDependencyInstallation,
	getPackageJSON,
} from '@namics/frontend-defaults-platform-core';

export type WebpackAnalyzerResult = {
	webpack: boolean;
	webpackInstallation?: DependencyInstallation;
	webpackConfigurations?: string[];
	webpackInstalledDependencies?: string[];
};

export const webpackAnalyzer = async (cwd: string): Promise<WebpackAnalyzerResult> => {
	const possibleWebpackConfigs = await findFilesByPattern(cwd, 'webpack.*.{js,ts}');

	if (possibleWebpackConfigs.length > 0) {
		const pkg = await getPackageJSON(cwd);
		const webpackInstallation = await getDependencyInstallation(cwd, 'webpack');
		const possibleWebpackDependencies = Object.keys(pkg.devDependencies || {}).filter(
			(name) => !!~name.indexOf('webpack')
		);

		return {
			webpack: true,
			webpackInstallation,
			webpackConfigurations: possibleWebpackConfigs,
			webpackInstalledDependencies: possibleWebpackDependencies,
		};
	}

	return {
		webpack: possibleWebpackConfigs.length > 0,
	};
};
