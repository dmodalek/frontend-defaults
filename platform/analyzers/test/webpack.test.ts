import { ProjectAnalyzer } from '@namics/frontend-defaults-platform-core';
import { getFixturePath } from './suite';
import { WebpackAnalyzer, WebpackAnalyzerResult } from '../src/webpack';

const FIXTURE = getFixturePath('webpack-project');

const getFixtureAnalyzer = async (context: string): Promise<ProjectAnalyzer<WebpackAnalyzerResult>> => {
	return new ProjectAnalyzer<WebpackAnalyzerResult>({
		context,
		analyzers: [WebpackAnalyzer],
	}).boot();
};

describe('Analyzers', () => {
	describe('WebpackAnalyzer', () => {
		it('should not crash the ProjectAnalyzer', async () => {
			expect(async () => await getFixtureAnalyzer(FIXTURE)).not.toThrow();
		});

		it('should analyze a project with installation correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE);

			expect(analyzer.analytics.webpack).toEqual(true);
			expect(analyzer.analytics.webpackInstallation).toEqual('4.29.0');
			expect(analyzer.analytics).toMatchSnapshot();
		});

		it('should analyze a project and find configs as well as plugins correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE);

			// ['webpack.config.js', 'webpack.prod.config.js']
			expect(analyzer.analytics.webpackConfigurations).toHaveLength(2);
			// ['webpack', 'html-webpack-plugin', 'ts-config-webpack-plugin']
			expect(analyzer.analytics.webpackInstalledDependencies).toHaveLength(3);
			expect(analyzer.analytics).toMatchSnapshot();
		});
	});
});
