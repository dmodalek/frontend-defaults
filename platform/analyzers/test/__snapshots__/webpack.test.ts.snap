// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers WebpackAnalyzer should analyze a project and find configs as well as plugins correctly 1`] = `
Object {
  "webpack": true,
  "webpackConfigurations": Array [
    "webpack.config.js",
    "webpack.prod.config.js",
  ],
  "webpackInstallation": Object {
    "declared": "4.29.0",
    "installed": null,
    "latest": Any<String>,
    "location": "devDependencies",
  },
  "webpackInstalledDependencies": Array [
    "webpack",
    "html-webpack-plugin",
    "ts-config-webpack-plugin",
  ],
}
`;

exports[`Analyzers WebpackAnalyzer should analyze a project with installation correctly 1`] = `
Object {
  "webpack": true,
  "webpackConfigurations": Array [
    "webpack.config.js",
    "webpack.prod.config.js",
  ],
  "webpackInstallation": Object {
    "declared": "4.29.0",
    "installed": null,
    "latest": Any<String>,
    "location": "devDependencies",
  },
  "webpackInstalledDependencies": Array [
    "webpack",
    "html-webpack-plugin",
    "ts-config-webpack-plugin",
  ],
}
`;
